server/src/main/java/io/pravega/schemaregistry/storage/impl/SchemaStoreImpl.java

```diff
@@ -130,14 +131,16 @@ public SchemaStoreImpl(Groups<T> groups, Schemas<T> schemas) {
     }
     
     @Override
-    public CompletableFuture<VersionInfo> addSchema(String namespace, String groupId, SchemaInfo schemaInfo, GroupProperties prop, Etag etag) {
-        return schemas.addSchema(schemaInfo, namespace, groupId)
-                .thenCompose(v -> getGroup(namespace, groupId).thenCompose(grp -> grp.addSchema(schemaInfo, prop, etag)));
+    public CompletableFuture<VersionInfo> addSchema(String namespace, String groupId, SchemaInfo schemaInfo, SchemaInfo normalized,
+                                                    BigInteger fingerprint, GroupProperties prop, Etag etag) {
+        // Store normalized form of schema with the global schemas while the original form is stored within the group.  
```

Comment:
Does it mean that we are storing each schema twice?

Reply:
yes, we store each schema in two separate tables - global schema table and group specific table. 
the global table is there to find all groups where a specific schema has been used. 

So its twice if its only used in one group, but its one additional times the number of groups its used in. 
So if 10 groups have the same schema, we have 11 copies of that schema stored in 11 different tables.

Reply:
And actually, in the global table, we could have just stored the fingerprint. 
But the reason i also keep the schema in the global table is, in case we decide to add APIs that allow users to see the global pool of schemas without any schema group context, then this would be helpful.

As an example, we could add APIs to list all global Schemas. 
Or search for schemas that have specific properties. it opens up for all those possibilities in future. 

currently we use that table for a single purpose of looking up all the groups where a specific schema is added to.

Reply:
Is it possible to have a single copy and replace with fingerprint copies and to the single copy? What would be the impact of doing it? 

I understand that each individual schema uses little space, but I don't really like the idea of keeping n copies unnecessarily.

Reply:
Conceptually we could keep a single copy (for example, just the global copy) and we generate a schema id. so the group could just store the id. 

The reason i kept the schemas in the group specific tables is so that querys for a group are optimized. 
Otherwise, for querying multiple schemas in a group, we first need to query group metadata and get schema ids.. and then query the global table to get the schemas for those ids. 

and secondly, with normalization, two schemas with different order of fields would be reduced to a single normalized copy. or we may have to store different ordered schemas as separate schemas. 
but that complicates the problem of ensuring that two schemas with identical normalized form need to be treated as identical and not registered within the group.

Reply:
> the reason i also keep the schema in the global table is, in case we decide to add APIs that allow users to see the global pool of schemas without any schema group context, then this would be helpful.

it makes sense to optimize for the common case, and frankly, I don't expect many duplicates of the same schema. I do feel that the global table is a waste, though. Even if we have calls to list schemas globally, those will be used rarely, so it makes sense to have their execution being more costly while making the overall system more efficient. I'm in favor of keeping a fingerprint and maybe pointers to schemas in the case of calls to globally list schemas, but of course, that depends on how much we gain by doing it. If storing fingerprint + point to schema is about the same size as the schema itself, then it is pointless.

Reply:
> so it makes sense to have their execution being more costly while making the overall system more efficient

I agree with the sentiment, but let us first evaluate what the cost of adding a schema to a global table really is. the global table is only accessed during a "new" schema addition which is a relatively rare event. 
And the cost is typically 2 KVT "get" calls followed by an "update" call. 
Latency cost: This will add few milliseconds to a new schema addition call. 
Storage cost: Two copies of schema being stored in two different pravega KVTs (global and group specific). If same schema is added to `n` groups, then we keep `n+1` copies of the schema binary. 

So the cost really is on storage. But that isnt too high - at the end of the day its stored in pravega tables, and occupies small space on LTS. 

It is important to note that global schemas table is not accessed in any group specific calls other than addSchema. 
And the get API that accesses this table is for finding all groups where a schema may have been added. 
So global schema does no add any cost on any group queries. 
it doesnt affect any subsequent queries on the group metadata.